src/main/java/com/shapesecurity/salvation2/Directives/HostSourceDirective.java

```diff
@@ -0,0 +1,207 @@
+package com.shapesecurity.salvation2.Directives;
+
+import com.shapesecurity.salvation2.Constants;
+import com.shapesecurity.salvation2.Directive;
+import com.shapesecurity.salvation2.Policy;
+import com.shapesecurity.salvation2.Values.Host;
+import com.shapesecurity.salvation2.Values.Scheme;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+public abstract class HostSourceDirective extends Directive {
+	protected List<Scheme> schemes = new ArrayList<>();
+	protected List<Host> hosts = new ArrayList<>();
+	protected boolean star = false;
+	protected boolean self = false;
+
+	protected String none = null;
```

Comment:
Why is this a string instead of a boolean?

Reply:
So that the input can be serialized without losing casing.

Reply:
I don't see what would permit `'none'` to have any other casing. The ABNF requires exactly `'none'`.

> A serialized source list is an ASCII string, consisting of a whitespace-delimited series of source expressions, adhering to the following ABNF grammar [RFC5234]:

Reply:
Strings in ABNF grammars are case-insensitive; read the [RFC](https://tools.ietf.org/html/rfc5234).

But also, as I keep telling you, the grammar is never actually enforced by anything. The thing that matters is the algorithms. And in [the algorithms](https://w3c.github.io/webappsec-csp/#match-url-to-source-list) it does ASCII case-insensitive comparisons for `'none'`.